crates/iceberg/src/arrow.rs

```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+        Ok(())
+    }
+
+    fn visit_binary(&mut self, predicate: &BinaryExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn visit_set(&mut self, predicate: &SetExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn bound_reference(&mut self, reference: &BoundReference) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+}
+
+struct PredicateConverter<'a> {
+    pub columns: &'a Vec<usize>,
+    pub projection_mask: ProjectionMask,
+    pub parquet_schema: &'a SchemaDescriptor,
+    pub column_map: &'a HashMap<i32, usize>,
+}
+
+fn get_arrow_datum(datum: &Datum) -> Box<dyn ArrowDatum> {
+    match datum.literal() {
+        PrimitiveLiteral::Boolean(value) => Box::new(BooleanArray::new_scalar(*value)),
+        PrimitiveLiteral::Int(value) => Box::new(Int32Array::new_scalar(*value)),
+        PrimitiveLiteral::Long(value) => Box::new(Int64Array::new_scalar(*value)),
+        PrimitiveLiteral::Float(value) => Box::new(Float32Array::new_scalar(value.as_f32())),
+        PrimitiveLiteral::Double(value) => Box::new(Float64Array::new_scalar(value.as_f64())),
+        _ => todo!("Unsupported literal type"),
+    }
+}
+
+impl<'a> BoundPredicateVisitor for PredicateConverter<'a> {
+    type T = Box<dyn ArrowPredicate>;
+    type U = usize;
+
+    fn visit_always_true(&mut self) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            |batch| Ok(BooleanArray::from(vec![true; batch.num_rows()])),
+        )))
+    }
+
+    fn visit_always_false(&mut self) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            |batch| Ok(BooleanArray::from(vec![false; batch.num_rows()])),
+        )))
+    }
+
+    fn visit_unary(&mut self, predicate: &UnaryExpression<BoundReference>) -> Result<Self::T> {
+        let term_index = self.bound_reference(predicate.term())?;
+
+        match predicate.op() {
+            PredicateOperator::IsNull => Ok(Box::new(ArrowPredicateFn::new(
+                self.projection_mask.clone(),
+                move |batch| {
+                    let column = batch.column(term_index);
+                    is_null(column)
+                },
+            ))),
+            PredicateOperator::NotNull => Ok(Box::new(ArrowPredicateFn::new(
+                self.projection_mask.clone(),
+                move |batch| {
+                    let column = batch.column(term_index);
+                    is_not_null(column)
+                },
+            ))),
+            PredicateOperator::IsNan => {
+                todo!("IsNan is not supported yet")
+            }
+            PredicateOperator::NotNan => {
+                todo!("NotNan is not supported yet")
+            }
```

Comment:
Seems there are no existing kernels for is_nan and not_nan operators. I've not added them in this PR and will implement them in following work.